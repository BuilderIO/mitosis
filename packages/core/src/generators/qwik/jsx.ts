import { MitosisNode } from '../../types/mitosis-node';
import { DIRECTIVES } from './directives';
import {
  File,
  INDENT,
  invoke,
  NL,
  SrcBuilder,
  quote,
  UNINDENT,
  lastProperty,
} from './src-generator';
import { CssStyles } from './styles';

export function renderJSXNodes(
  file: File,
  directives: Map<string, string>,
  handlers: Map<string, string>,
  children: MitosisNode[],
  styles: Map<string, CssStyles>,
  parentSymbolBindings: Record<string, string>,
  root = true,
): any {
  return function(this: SrcBuilder) {
    if (children.length == 0) return;
    if (root) this.emit('(', INDENT, NL);
    const needsFragment = root && children.length > 1;
    file.import(file.qwikModule, 'h');
    if (needsFragment) {
      file.import(file.qwikModule, 'Fragment');
      this.jsxBeginFragment(file.import(file.qwikModule, 'Fragment'));
    }
    children.forEach((child) => {
      if (isEmptyTextNode(child)) return;
      if (isTextNode(child)) {
        if (child.bindings._text !== undefined) {
          this.jsxTextBinding(child.bindings._text);
        } else {
          this.isJSX
            ? this.emit(child.properties._text)
            : this.jsxTextBinding(quote(child.properties._text!));
        }
      } else {
        let childName = child.name;
        const directive = DIRECTIVES[childName];
        if (typeof directive == 'function') {
          this.emit(
            directive(child, () =>
              renderJSXNodes(
                file,
                directives,
                handlers,
                child.children,
                styles,
                {},
                false,
              ).call(this),
            ),
          );
        } else {
          if (typeof directive == 'string') {
            directives.set(childName, directive);
            if (file.module !== 'med') {
              file.import('./med.js', childName);
            }
          }
          if (isSymbol(childName)) {
            // TODO(misko): We are hard coding './med.js' which is not right.
            file.import('./med.js', childName);
            let exportedChildName = file.exports.get(childName);
            if (exportedChildName) {
              childName = exportedChildName;
            }
          }
          let props: Record<string, any> = child.properties;
          const css = child.bindings.css;
          if (css) {
            props = { ...props };
            props.class = addClass(styles.get(css)!.CLASS_NAME, props.class);
          }
          const symbolBindings: Record<string, string> = {};
          const bindings = rewriteHandlers(
            file,
            handlers,
            child.bindings,
            symbolBindings,
          );
          this.jsxBegin(childName, props, {
            ...bindings,
            ...parentSymbolBindings,
          });
          renderJSXNodes(
            file,
            directives,
            handlers,
            child.children,
            styles,
            symbolBindings,
            false,
          ).call(this);
          this.jsxEnd(childName);
        }
      }
    });
    if (needsFragment) {
      this.jsxEndFragment();
    }
    if (root) this.emit(UNINDENT, ')');
  };
}

function isSymbol(name: string): boolean {
  return name.charAt(0) == name.charAt(0).toUpperCase();
}

function addClass(
  className: string,
  existingClass: string | undefined,
): string {
  return [className, ...(existingClass ? existingClass.split(' ') : [])].join(
    ' ',
  );
}

function isEmptyTextNode(child: MitosisNode) {
  return child.properties._text?.trim() == '';
}

function isTextNode(child: MitosisNode) {
  return (
    child.properties._text !== undefined || child.bindings._text !== undefined
  );
}

/**
 * Rewrites bindings:
 * - Remove `css`
 * - Rewrites event handles
 * - Extracts symbol bindings.
 *
 * @param file
 * @param handlers
 * @param bindings
 * @param symbolBindings Options record which will receive the symbol bindings
 * @returns
 */
function rewriteHandlers(
  file: File,
  handlers: Map<string, string>,
  bindings: Record<string, string | undefined>,
  symbolBindings: Record<string, string>,
): Record<string, string> {
  const outBindings: Record<string, string> = {};
  for (let key in bindings) {
    if (Object.prototype.hasOwnProperty.call(bindings, key)) {
      let binding: any = bindings[key]!;
      let handlerBlock: string | undefined;
      if (binding != null) {
        if (key == 'css') {
          continue;
        } else if ((handlerBlock = handlers.get(binding))) {
          key = `on:${key.substring(2).toLowerCase()}`;
          binding = invoke(file.import(file.qwikModule, 'qrl'), [
            quote(file.qrlPrefix + 'high.js'),
            quote(handlerBlock),
            '[__props__, __state__]',
          ]);
        } else if (symbolBindings && key.startsWith('symbol.data.')) {
          symbolBindings[lastProperty(key)] = binding;
        }
        outBindings[key] = binding;
      }
    }
  }
  return outBindings;
}
