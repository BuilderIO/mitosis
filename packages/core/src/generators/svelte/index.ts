export { componentToSvelte } from './svelte';
export * from './types';
