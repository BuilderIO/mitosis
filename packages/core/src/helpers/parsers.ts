import * as babel from '@babel/core';

const jsxPlugin = require('@babel/plugin-syntax-jsx');
const tsPreset = require('@babel/preset-typescript');
const decorators = require('@babel/plugin-syntax-decorators');

export function parseCode(code: string) {
  const ast = babel.parse(code, {
    presets: [[tsPreset, { isTSX: true, allExtensions: true }]],
    plugins: [[decorators, { legacy: true }], jsxPlugin],
  });
  const body = babel.types.isFile(ast)
    ? ast.program.body
    : babel.types.isProgram(ast)
    ? ast.body
    : [];
  return body;
}

export const isCodeBodyExpression = (body: babel.types.Statement[]) =>
  body.length == 1 &&
  (babel.types.isExpression(body[0]) ||
    babel.types.isExpressionStatement(body[0]));

/**
 * Returns `true` if the `code` is a valid expression. (vs a statement)
 */
export function isExpression(code: string): boolean {
  try {
    const body = parseCode(code);
    return isCodeBodyExpression(body);
  } catch (e) {
    return false;
  }
}

export const isCodeBodyIdentifier = (body: babel.types.Statement[]) =>
  body.length == 1 && babel.types.isIdentifier(body[0]);
