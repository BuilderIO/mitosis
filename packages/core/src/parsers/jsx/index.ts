export * from './jsx';
export { selfClosingTags } from './helpers';
export { METADATA_HOOK_NAME } from './hooks';
