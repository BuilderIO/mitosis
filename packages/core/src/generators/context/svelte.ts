import { format } from 'prettier/standalone';
import { getMemberObjectString } from '../../helpers/get-state-object-string';
import { MitosisContext } from '../../types/mitosis-context';
import { BaseTranspilerOptions } from '../../types/config';

interface ContextToSvelteOptions
  extends Pick<BaseTranspilerOptions, 'prettier'> {}

export const contextToSvelte =
  (options: ContextToSvelteOptions = {}) =>
  ({ context }: { context: MitosisContext }): string => {
    let str = `
  const key = Symbol();  

  export default {
    ${context.name}: ${getMemberObjectString(context.value)}, 
    key 
  }
  `;

    if (options.prettier !== false) {
      try {
        str = format(str, {
          parser: 'typescript',
          plugins: [
            require('prettier/parser-typescript'), // To support running in browsers
          ],
        });
      } catch (err) {
        console.error('Format error for file:', str);
        throw err;
      }
    }

    return str;
  };
