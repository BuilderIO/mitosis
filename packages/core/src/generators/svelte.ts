import dedent from 'dedent';
import { format } from 'prettier/standalone';
import traverse from 'traverse';
import { collectCss } from '../helpers/collect-styles';
import { fastClone } from '../helpers/fast-clone';
import { getProps } from '../helpers/get-props';
import { getRefs } from '../helpers/get-refs';
import { getStateObjectStringFromComponent } from '../helpers/get-state-object-string';
import { isMitosisNode } from '../helpers/is-mitosis-node';
import { renderPreComponent } from '../helpers/render-imports';
import { stripStateAndPropsRefs } from '../helpers/strip-state-and-props-refs';
import { selfClosingTags } from '../parsers/jsx';
import { MitosisComponent } from '../types/mitosis-component';
import { MitosisNode } from '../types/mitosis-node';
import {
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../modules/plugins';
import isChildren from '../helpers/is-children';
import { stripMetaProperties } from '../helpers/strip-meta-properties';
import { removeSurroundingBlock } from '../helpers/remove-surrounding-block';
import { BaseTranspilerOptions, Transpiler } from '../types/config';
import { gettersToFunctions } from '../helpers/getters-to-functions';
import { babelTransformCode } from '../helpers/babel-transform';

import { pipe } from 'fp-ts/lib/function';
export interface ToSvelteOptions extends BaseTranspilerOptions {
  stateType?: 'proxies' | 'variables';
}

const mappers: {
  [key: string]: (json: MitosisNode, options: ToSvelteOptions) => string;
} = {
  Fragment: (json, options) => {
    if (json.bindings.innerHTML) {
      return BINDINGS_MAPPER.innerHTML(json, options);
    } else if (json.children.length > 0) {
      return `${json.children
        .map((item) => blockToSvelte(item, options))
        .join('\n')}`;
    } else {
      return '';
    }
  },
};

const BINDINGS_MAPPER = {
  innerHTML: (json: MitosisNode, options: ToSvelteOptions) =>
    `{@html ${stripStateAndPropsRefs(json.bindings.innerHTML)}}`,
};

export const blockToSvelte = (
  json: MitosisNode,
  options: ToSvelteOptions,
): string => {
  if (mappers[json.name]) {
    return mappers[json.name](json, options);
  }

  if (isChildren(json)) {
    return `<slot></slot>`;
  }

  if (json.properties._text) {
    return json.properties._text;
  }

  if (json.bindings._text) {
    return `{${stripStateAndPropsRefs(json.bindings._text as string, {
      includeState: options.stateType === 'variables',
    })}}`;
  }

  let str = '';

  if (json.name === 'For') {
    str += `{#each ${stripStateAndPropsRefs(json.bindings.each as string, {
      includeState: options.stateType === 'variables',
    })} as ${json.properties._forName}, index }`;
    str += json.children.map((item) => blockToSvelte(item, options)).join('\n');
    str += `{/each}`;
  } else if (json.name === 'Show') {
    str += `{#if ${stripStateAndPropsRefs(json.bindings.when as string, {
      includeState: options.stateType === 'variables',
    })} }`;
    str += json.children.map((item) => blockToSvelte(item, options)).join('\n');
    str += `{/if}`;
  } else {
    str += `<${json.name} `;

    if (json.bindings._spread) {
      str += `{...${stripStateAndPropsRefs(json.bindings._spread as string, {
        includeState: options.stateType === 'variables',
      })}}`;
    }

    for (const key in json.properties) {
      const value = json.properties[key];
      str += ` ${key}="${value}" `;
    }
    for (const key in json.bindings) {
      if (key === 'innerHTML') {
        continue;
      }
      if (key === '_spread') {
        continue;
      }
      const value = json.bindings[key] as string;
      // TODO: proper babel transform to replace. Util for this
      const useValue = stripStateAndPropsRefs(value, {
        includeState: options.stateType === 'variables',
      });

      if (key.startsWith('on')) {
        const event = key.replace('on', '').toLowerCase();
        // TODO: handle quotes in event handler values
        str += ` on:${event}="{event => ${removeSurroundingBlock(useValue)}}" `;
      } else if (key === 'ref') {
        str += ` bind:this={${useValue}} `;
      } else {
        str += ` ${key}={${useValue}} `;
      }
    }
    // if we have innerHTML, it doesn't matter whether we have closing tags or not, or children or not.
    // we use the innerHTML content as children and don't render the self-closing tag.
    if (json.bindings.innerHTML) {
      str += '>';
      str += BINDINGS_MAPPER.innerHTML(json, options);
      str += `</${json.name}>`;
      return str;
    }

    if (selfClosingTags.has(json.name)) {
      return str + ' />';
    }
    str += '>';
    if (json.children) {
      str += json.children
        .map((item) => blockToSvelte(item, options))
        .join('\n');
    }

    str += `</${json.name}>`;
  }
  return str;
};

/**
 * Replace
 *    <input value={state.name} onChange={event => state.name = event.target.value}
 * with
 *    <input bind:value={state.name}/>
 * when easily identified, for more idiomatic svelte code
 */
const useBindValue = (json: MitosisComponent, options: ToSvelteOptions) => {
  traverse(json).forEach(function (item) {
    if (isMitosisNode(item)) {
      const { value, onChange } = item.bindings;
      if (value && onChange) {
        if (
          (onChange as string).replace(/\s+/g, '') ===
          `${value}=event.target.value`
        ) {
          delete item.bindings.value;
          delete item.bindings.onChange;
          item.bindings['bind:value'] = value;
        }
      }
    }
  });
};

export const componentToSvelte =
  (options: ToSvelteOptions = {}): Transpiler =>
  ({ component }) => {
    const useOptions: ToSvelteOptions = {
      ...options,
      stateType: 'variables',
    };
    // Make a copy we can safely mutate, similar to babel's toolchain
    let json = fastClone(component);
    if (useOptions.plugins) {
      json = runPreJsonPlugins(json, useOptions.plugins);
    }

    const refs = Array.from(getRefs(json));
    useBindValue(json, useOptions);

    gettersToFunctions(json);

    if (useOptions.plugins) {
      json = runPostJsonPlugins(json, useOptions.plugins);
    }
    const css = collectCss(json);
    stripMetaProperties(json);

    const dataString = pipe(
      getStateObjectStringFromComponent(json, {
        data: true,
        functions: false,
        getters: false,
        format: useOptions.stateType === 'proxies' ? 'object' : 'variables',
        keyPrefix: useOptions.stateType === 'variables' ? 'let ' : '',
        valueMapper: (code) =>
          stripStateAndPropsRefs(code, {
            includeState: useOptions.stateType === 'variables',
          }),
      }),
      babelTransformCode,
    );

    const getterString = pipe(
      getStateObjectStringFromComponent(json, {
        data: false,
        getters: true,
        functions: false,
        format: 'variables',
        keyPrefix: '$: ',
        valueMapper: (code) =>
          stripStateAndPropsRefs(
            code
              .replace(/^get ([a-zA-Z_\$0-9]+)/, '$1 = ')
              .replace(/\)/, ') => '),
            {
              includeState: useOptions.stateType === 'variables',
            },
          ),
      }),
      babelTransformCode,
    );

    const functionsString = pipe(
      getStateObjectStringFromComponent(json, {
        data: false,
        getters: false,
        functions: true,
        format: 'variables',
        keyPrefix: 'function ',
        valueMapper: (code) =>
          stripStateAndPropsRefs(code, {
            includeState: useOptions.stateType === 'variables',
          }),
      }),
      babelTransformCode,
    );

    const hasData = dataString.length > 4;

    const props = Array.from(getProps(json));

    let str = dedent`
    <script>
      ${!json.hooks.onMount?.code ? '' : `import { onMount } from 'svelte'`}
      ${
        !json.hooks.onUpdate?.length
          ? ''
          : `import { afterUpdate } from 'svelte'`
      }
      ${!json.hooks.onUnMount?.code ? '' : `import { onDestroy } from 'svelte'`}
      ${renderPreComponent(json, 'svelte')}

      ${
        !hasData || useOptions.stateType === 'variables'
          ? ''
          : `import onChange from 'on-change'`
      }
      ${refs
        .concat(props)
        .map((name) => `export let ${name};`)
        .join('\n')}

      ${functionsString.length < 4 ? '' : functionsString}
      ${getterString.length < 4 ? '' : getterString}

      ${
        useOptions.stateType === 'proxies'
          ? dataString.length < 4
            ? ''
            : `let state = onChange(${dataString}, () => state = state)`
          : dataString
      }

      ${
        !json.hooks.onMount?.code
          ? ''
          : `onMount(() => { ${stripStateAndPropsRefs(json.hooks.onMount.code, {
              includeState: useOptions.stateType === 'variables',
            })} });`
      }

      ${
        !json.hooks.onUpdate?.length
          ? ''
          : json.hooks.onUpdate
              .map(
                (hook) =>
                  `afterUpdate(() => { ${stripStateAndPropsRefs(hook.code, {
                    includeState: useOptions.stateType === 'variables',
                  })} })`,
              )
              .join(';')
      }

      ${
        !json.hooks.onUnMount?.code
          ? ''
          : `onDestroy(() => { ${stripStateAndPropsRefs(
              json.hooks.onUnMount.code,
              {
                includeState: useOptions.stateType === 'variables',
              },
            )} });`
      }
    </script>

    ${json.children.map((item) => blockToSvelte(item, useOptions)).join('\n')}

    ${
      !css.trim().length
        ? ''
        : `<style>
      ${css}
    </style>`
    }
  `;

    if (useOptions.plugins) {
      str = runPreCodePlugins(str, useOptions.plugins);
    }
    if (useOptions.prettier !== false) {
      try {
        str = format(str, {
          parser: 'svelte',
          plugins: [
            // To support running in browsers
            require('prettier/parser-html'),
            require('prettier/parser-postcss'),
            require('prettier/parser-babel'),
            require('prettier/parser-typescript'),
            require('prettier-plugin-svelte'),
          ],
        });
      } catch (err) {
        console.warn('Could not prettify', { string: str }, err);
      }
    }
    if (useOptions.plugins) {
      str = runPostCodePlugins(str, useOptions.plugins);
    }
    return str;
  };
