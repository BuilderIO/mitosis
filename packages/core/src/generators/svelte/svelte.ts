import dedent from 'dedent';
import { format } from 'prettier/standalone';
import traverse from 'traverse';
import { collectCss } from '../../helpers/styles/collect-css';
import { hasStyle } from '../../helpers/styles/helpers';
import { fastClone } from '../../helpers/fast-clone';
import { getProps } from '../../helpers/get-props';
import { getRefs } from '../../helpers/get-refs';
import {
  stringifyContextValue,
  getStateObjectStringFromComponent,
} from '../../helpers/get-state-object-string';
import { isMitosisNode } from '../../helpers/is-mitosis-node';
import { renderPreComponent } from '../../helpers/render-imports';
import { MitosisComponent } from '../../types/mitosis-component';
import {
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../../modules/plugins';
import { stripMetaProperties } from '../../helpers/strip-meta-properties';
import { TranspilerGenerator } from '../../types/transpiler';
import { gettersToFunctions } from '../../helpers/getters-to-functions';
import { babelTransformCode } from '../../helpers/babel-transform';
import { flow, pipe } from 'fp-ts/lib/function';
import { hasContext } from '../helpers/context';
import { isSlotProperty } from '../../helpers/slots';
import json5 from 'json5';
import { FUNCTION_HACK_PLUGIN } from '../helpers/functions';
import { mergeOptions } from '../../helpers/merge-options';
import { CODE_PROCESSOR_PLUGIN } from '../../helpers/plugins/process-code';
import { stripStateAndProps } from './helpers';
import { ToSvelteOptions } from './types';
import { blockToSvelte } from './blocks';
import { stripGetter } from '../../helpers/patterns';

const getContextCode = (json: MitosisComponent) => {
  const contextGetters = json.context.get;
  return Object.entries(contextGetters)
    .map(([key, context]): string => {
      const { name } = context;

      return `let ${key} = getContext(${name}.key);`;
    })
    .join('\n');
};

const setContextCode = ({
  json,
  options,
}: {
  json: MitosisComponent;
  options: ToSvelteOptions;
}) => {
  const processCode = stripStateAndProps({ json, options });

  return Object.values(json.context.set)
    .map((context) => {
      const { value, name, ref } = context;
      const key = value ? `${name}.key` : name;

      const valueStr = value
        ? processCode(stringifyContextValue(value))
        : ref
        ? processCode(ref)
        : 'undefined';

      return `setContext(${key}, ${valueStr});`;
    })
    .join('\n');
};

/**
 * Replace
 *    <input value={state.name} onChange={event => state.name = event.target.value}
 * with
 *    <input bind:value={state.name}/>
 * when easily identified, for more idiomatic svelte code
 */
const useBindValue = (json: MitosisComponent, options: ToSvelteOptions) => {
  function normalizeStr(str: string) {
    return str
      .trim()
      .replace(/\n|\r/g, '')
      .replace(/^{/, '')
      .replace(/}$/, '')
      .replace(/;$/, '')
      .replace(/\s+/g, '');
  }
  traverse(json).forEach(function (item) {
    if (isMitosisNode(item)) {
      const { value, onChange } = item.bindings;
      if (value && onChange) {
        const { arguments: cusArgs = ['event'] } = onChange;
        if (
          normalizeStr(onChange.code) === `${normalizeStr(value.code)}=${cusArgs[0]}.target.value`
        ) {
          delete item.bindings.value;
          delete item.bindings.onChange;
          item.bindings['bind:value'] = value;
        }
      }
    }
  });
};

const DEFAULT_OPTIONS: ToSvelteOptions = {
  stateType: 'variables',
  prettier: true,
  plugins: [FUNCTION_HACK_PLUGIN],
};

export const componentToSvelte: TranspilerGenerator<ToSvelteOptions> =
  (userProvidedOptions) =>
  ({ component }) => {
    const options = mergeOptions(DEFAULT_OPTIONS, userProvidedOptions);

    options.plugins = [
      ...(options.plugins || []),
      CODE_PROCESSOR_PLUGIN((codeType) => {
        switch (codeType) {
          case 'hooks':
            return flow(stripStateAndProps({ json, options }), babelTransformCode);
          case 'bindings':
          case 'hooks-deps':
          case 'state':
            return flow(stripStateAndProps({ json, options }), stripGetter);
          case 'properties':
            return stripStateAndProps({ json, options });
        }
      }),
    ];

    // Make a copy we can safely mutate, similar to babel's toolchain
    let json = fastClone(component);
    json = runPreJsonPlugins(json, options.plugins);

    const refs = Array.from(getRefs(json));
    useBindValue(json, options);

    gettersToFunctions(json);

    const props = Array.from(getProps(json)).filter((prop) => !isSlotProperty(prop));

    json = runPostJsonPlugins(json, options.plugins);

    const css = collectCss(json);
    stripMetaProperties(json);

    const dataString = pipe(
      getStateObjectStringFromComponent(json, {
        data: true,
        functions: false,
        getters: false,
        format: options.stateType === 'proxies' ? 'object' : 'variables',
        keyPrefix: options.stateType === 'variables' ? 'let ' : '',
      }),
      babelTransformCode,
    );

    const getterString = pipe(
      getStateObjectStringFromComponent(json, {
        data: false,
        getters: true,
        functions: false,
        format: 'variables',
        keyPrefix: '$: ',
        valueMapper: (code) => {
          return code
            .trim()
            .replace(/^([a-zA-Z_\$0-9]+)/, '$1 = ')
            .replace(/\)/, ') => ');
        },
      }),
      babelTransformCode,
    );

    const functionsString = pipe(
      getStateObjectStringFromComponent(json, {
        data: false,
        getters: false,
        functions: true,
        format: 'variables',
      }),
      babelTransformCode,
    );

    const hasData = dataString.length > 4;

    let str = '';

    const tsLangAttribute = options.typescript ? `lang='ts'` : '';

    if (options.typescript && json.types?.length) {
      str += dedent`
      <script context='module' ${tsLangAttribute}>
        ${json.types ? json.types.join('\n\n') + '\n' : ''}
      </script>
      \n
      \n
      `;
    }

    // prepare svelte imports
    let svelteImports: string[] = [];
    let svelteStoreImports: string[] = [];

    if (json.hooks.onMount?.code?.length) {
      svelteImports.push('onMount');
    }
    if (json.hooks.onUpdate?.length) {
      svelteImports.push('afterUpdate');
    }
    if (json.hooks.onUnMount?.code?.length) {
      svelteImports.push('onDestroy');
    }
    if (hasContext(component)) {
      svelteImports.push('getContext', 'setContext');
    }

    str += dedent`
      <script ${tsLangAttribute}>
      ${!svelteImports.length ? '' : `import { ${svelteImports.sort().join(', ')} } from 'svelte'`}
      ${
        !svelteStoreImports.length
          ? ''
          : `import { ${svelteStoreImports.sort().join(', ')} } from 'svelte/store'`
      }

      ${renderPreComponent({ component: json, target: 'svelte' })}

      ${!hasData || options.stateType === 'variables' ? '' : `import onChange from 'on-change'`}
      ${props
        .map((name) => {
          if (name === 'children') {
            return '';
          }

          let propDeclaration = `export let ${name}`;

          if (options.typescript && json.propsTypeRef && json.propsTypeRef !== 'any') {
            propDeclaration += `: ${json.propsTypeRef.split(' |')[0]}['${name}']`;
          }

          if (json.defaultProps && json.defaultProps.hasOwnProperty(name)) {
            propDeclaration += `=${json5.stringify(json.defaultProps[name])}`;
          }

          propDeclaration += ';';

          return propDeclaration;
        })
        .join('\n')}
      ${
        // https://svelte.dev/repl/bd9b56891f04414982517bbd10c52c82?version=3.31.0
        hasStyle(json)
          ? `
        function mitosis_styling (node, vars) {
          Object.entries(vars || {}).forEach(([ p, v ]) => {
            if (p.startsWith('--')) {
              node.style.setProperty(p, v);
            } else {
              node.style[p] = v;
            }
          })
        }
      `
          : ''
      }
      ${getContextCode(json)}
      ${setContextCode({ json, options })}

      ${functionsString.length < 4 ? '' : functionsString}
      ${getterString.length < 4 ? '' : getterString}

      ${refs.map((ref) => `let ${stripStateAndProps({ json, options })(ref)}`).join('\n')}

      ${
        options.stateType === 'proxies'
          ? dataString.length < 4
            ? ''
            : `let state = onChange(${dataString}, () => state = state)`
          : dataString
      }
      ${json.hooks.onInit?.code ?? ''}
      
      ${!json.hooks.onMount?.code ? '' : `onMount(() => { ${json.hooks.onMount.code} });`}

      ${
        json.hooks.onUpdate
          ?.map(({ code, deps }, index) => {
            if (!deps) {
              return `afterUpdate(() => { ${code} });`;
            }

            const fnName = `onUpdateFn_${index}`;
            return `
              function ${fnName}() {
                ${code}
              }
              $: ${fnName}(...${deps})
            `;
          })
          .join(';') || ''
      }

      ${!json.hooks.onUnMount?.code ? '' : `onDestroy(() => { ${json.hooks.onUnMount.code} });`}
    </script>

    ${json.children
      .map((item) =>
        blockToSvelte({
          json: item,
          options: options,
          parentComponent: json,
        }),
      )
      .join('\n')}

    ${
      !css.trim().length
        ? ''
        : `<style>
      ${css}
    </style>`
    }
  `;

    str = runPreCodePlugins(str, options.plugins);

    if (options.prettier !== false) {
      try {
        str = format(str, {
          parser: 'svelte',
          plugins: [
            // To support running in browsers
            require('prettier/parser-html'),
            require('prettier/parser-postcss'),
            require('prettier/parser-babel'),
            require('prettier/parser-typescript'),
            require('prettier-plugin-svelte'),
          ],
        });
      } catch (err) {
        console.warn('Could not prettify');
        console.warn({ string: str }, err);
      }
    }
    str = runPostCodePlugins(str, options.plugins);

    return str;
  };
