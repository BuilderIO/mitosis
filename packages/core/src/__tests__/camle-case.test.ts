import MyComponent from './data/map-camle-cased-attributes.raw?raw'
import { parseJsx } from '../parsers/jsx';
import { componentToSvelte } from '../generators/svelte';

describe('map camle cased', () => {
    test('map camle cased', () => {
        const component = parseJsx(MyComponent);
        const svelteComponent = componentToSvelte()({component})
        expect(svelteComponent).toMatchSnapshot();
    })
})