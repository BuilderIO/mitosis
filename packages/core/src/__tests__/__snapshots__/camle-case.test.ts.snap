// Vitest Snapshot v1

exports[`map camle cased > map camle cased 1`] = `
"<script>
  import React from \\"@builder.io/react\\";

  let name = \\"Steve\\";
</script>

<div>
  <img srcSet=\\"http://example.com\\" />
</div>"
`;
